Fix hang loop when scan a camera device with no support match v4l2_fmt_maps array

iysheng opened the pull request:
when I scan camera devices , I found it will hang loop when scan a camera device with no support match v4l2_fmt_maps array, like
![Screenshot from 2025-05-20 01-03-14](https://github.com/user-attachments/assets/09e0face-864f-4bcc-abc0-473b583a8b65)

sauwming commented:
The proposed fix is not quite correct. The loop is relying on `xioctl()` to return non-success when it doesn't return a valid pixel format, but it seems that this is not the case.

So perhaps it should be:
```
if (status != PJ_SUCCESS || fdesc.pixelformat ==0)
    break;
```

nanangizz commented:
Perhaps also add a bit more safety net here (for other xioctl() usages):

```
/* util: ioctl that tries harder. */
static pj_status_t xioctl(int fh, int request, void *arg)
{
    enum { RETRY = MAX_IOCTL_RETRY };
    int r, c=0;

    do {
        r = v4l2_ioctl(fh, request, arg);
    } while (r==-1 && c++<RETRY && ((errno==EINTR) || (errno==EAGAIN)));

    // ==== CHANGES START ====
    if (r == -1) {
        pj_status_t status = pj_get_os_error();
        return (status==PJ_SUCCESS)? PJ_EUNKOWN : status;
    }

    return PJ_SUCCESS;
    // ==== CHANGES END ====
}
```

iysheng commented:
> Perhaps also add a bit more safety net here (for other xioctl() usages):
> 
> ```
> /* util: ioctl that tries harder. */
> static pj_status_t xioctl(int fh, int request, void *arg)
> {
>     enum { RETRY = MAX_IOCTL_RETRY };
>     int r, c=0;
> 
>     do {
>         r = v4l2_ioctl(fh, request, arg);
>     } while (r==-1 && c++<RETRY && ((errno==EINTR) || (errno==EAGAIN)));
> 
>     // ==== CHANGES START ====
>     if (r == -1) {
>         pj_status_t status = pj_get_os_error();
>         return (status==PJ_SUCCESS)? PJ_EUNKOWN : status;
>     }
> 
>     return PJ_SUCCESS;
>     // ==== CHANGES END ====
> }
> ```

I prefer this way to decide the return value, this makes these releate code more clearner. Besides, I wonder why check the j value whether overflow is inappropriate here?

iysheng commented:
> The proposed fix is not quite correct. The loop is relying on `xioctl()` to return non-success when it doesn't return a valid pixel format, but it seems that this is not the case.
> 
> So perhaps it should be:
> 
> ```
> if (status != PJ_SUCCESS || fdesc.pixelformat ==0)
>     break;
> ```

looks good, but  I though compare ``fdesc.pixelformat ==0`` maybe not suite here, looks a little imprecisely here :sweat_smile: :.

sauwming commented:
> I prefer this way to decide the return value, this makes these releate code more clearner. Besides, I wonder why check the j value whether overflow is inappropriate here?

j is not used as an index to access the array, so it can't overflow (in other words, it is valid for j to be larger than `PJ_ARRAY_SIZE(fmt_cap)`, as long as `fmt_cnt` is smaller than `PJ_ARRAY_SIZE(fmt_cap)`).

iysheng commented:
> > I prefer this way to decide the return value, this makes these releate code more clearner. Besides, I wonder why check the j value whether overflow is inappropriate here?
> 
> j is not used as an index to access the array, so it can't overflow (in other words, it is valid for j to be larger than `PJ_ARRAY_SIZE(fmt_cap)`, as long as `fmt_cnt` is smaller than `PJ_ARRAY_SIZE(fmt_cap)`).

ok, thanks for your patient relpy. After I add debug print with the r value, and I get log like
```
[root@M3562:/tmp/aarch64_libs]$ ./vid_streamutil 1 1
15:19:20.874         os_core_unix.c !pjlib 2.15-dev for POSIX initialized
15:19:20.875         sip_endpoint.c  .Creating endpoint instance...
15:19:20.882                  pjlib  .select() I/O Queue created (0x5580875348)
15:19:20.882         sip_endpoint.c  .Module "mod-msg-print" registered
15:19:20.882        sip_transport.c  .Transport manager created.
15:19:20.882           pjsua_core.c  .PJSUA state changed: NULL --> CREATED
15:19:20.882       vid_streamutil.c  Initializing stack..
15:19:20.882         sip_endpoint.c  .Module "mod-pjsua-log" registered
15:19:20.882         sip_endpoint.c  .Module "mod-tsx-layer" registered
15:19:20.882         sip_endpoint.c  .Module "mod-stateful-util" registered
15:19:20.882         sip_endpoint.c  .Module "mod-ua" registered
15:19:20.882         sip_endpoint.c  .Module "mod-100rel" registered
15:19:20.882         sip_endpoint.c  .Module "mod-pjsua" registered
15:19:20.882         sip_endpoint.c  .Module "mod-invite" registered
15:19:20.929             alsa_dev.c  ..Added sound device samplerate
15:19:20.933             alsa_dev.c  ..Added sound device speexrate
15:19:20.934             alsa_dev.c  ..Added sound device oss
15:19:21.171             alsa_dev.c  ..Added sound device pulse
15:19:21.175             alsa_dev.c  ..Added sound device upmix
15:19:21.179             alsa_dev.c  ..Added sound device vdownmix
15:19:21.183             alsa_dev.c  ..Added sound device default
15:19:21.191             alsa_dev.c  ..Added sound device sysdefault:CARD=rockchiprk809co
15:19:21.201             alsa_dev.c  ..ALSA driver found 8 devices
15:19:21.201             alsa_dev.c  ..ALSA initialized
15:19:21.201                  pjlib  ..select() I/O Queue created (0x55808bdd98)
15:19:21.208           conference.c  ..Creating conference bridge with 254 ports
15:19:21.208           Master/sound  ..Using delay buffer with WSOLA.
15:19:21.233            pjsua_vid.c  ..Initializing video subsystem..
15:19:21.233             vid_conf.c  ...Created video conference bridge with 32 ports
15:19:21.240            vid_codec.c  ...codeid:H264/97
15:19:21.240           openh264.cpp  ...OpenH264 codec initialized
15:19:21.241            vid_codec.c  ...codeid:H263-1998/96
15:19:21.241            vid_codec.c  ...codeid:H263/34
15:19:21.241            vid_codec.c  ...codeid:H261/31
15:19:21.241            vid_codec.c  ...codeid:JPEG/26
15:19:21.241            vid_codec.c  ...codeid:MP4V/0
15:19:21.241            vid_codec.c  ...codeid:VP8/102
15:19:21.241                  vpx.c  ...VPX codec initialized
Using mplane plugin for capture
15:19:21.242             v4l2_dev.c  ... r=0
15:19:21.242             v4l2_dev.c  ...Found capture device rkisp_mainpath
15:19:21.242             v4l2_dev.c  ...  Enumerating formats:
15:19:21.242             v4l2_dev.c  ... r=0
15:19:21.242             v4l2_dev.c  ...   Supported: UYVY 4:2:2
15:19:21.242             v4l2_dev.c  ... r=0
15:19:21.242             v4l2_dev.c  ...   Unsupported: Y/CbCr 4:2:2
15:19:21.242             v4l2_dev.c  ... r=0
15:19:21.242             v4l2_dev.c  ...   Unsupported: Y/CrCb 4:2:2
15:19:21.242             v4l2_dev.c  ... r=0
15:19:21.242             v4l2_dev.c  ...   Unsupported: Y/CrCb 4:2:0
15:19:21.242             v4l2_dev.c  ... r=0
15:19:21.242             v4l2_dev.c  ...   Unsupported: Y/CbCr 4:2:0
15:19:21.242             v4l2_dev.c  ... r=-22
15:19:21.242             v4l2_dev.c  ...   Unsupported:
15:19:21.242             v4l2_dev.c  ... r=-22
15:19:21.242             v4l2_dev.c  ...   Unsupported:
15:19:21.242             v4l2_dev.c  ... r=-22
15:19:21.242             v4l2_dev.c  ...   Unsupported:
15:19:21.242             v4l2_dev.c  ... r=0
15:19:21.242             v4l2_dev.c  ... r=0
15:19:21.242             v4l2_dev.c  ... r=0
15:19:21.242             v4l2_dev.c  ... r=0
15:19:21.242             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ... r=0
Using mplane plugin for capture
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ...Found capture device rkisp_selfpath
15:19:21.243             v4l2_dev.c  ...  Enumerating formats:
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ...   Supported: UYVY 4:2:2
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ...   Unsupported: Y/CbCr 4:2:2
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ...   Unsupported: Y/CrCb 4:2:2
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.243             v4l2_dev.c  ...   Unsupported: Y/CrCb 4:2:0
15:19:21.243             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ...   Unsupported: Y/CbCr 4:2:0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ...   Unsupported: 8-bit Greyscale
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ...   Unsupported: 32-bit BGRX 8-8-8-8
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ...   Unsupported: 16-bit RGB 5-6-5
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
Using mplane plugin for capture
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ...Found capture device rkisp_iqtool
15:19:21.244             v4l2_dev.c  ...  Enumerating formats:
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ...   Supported: UYVY 4:2:2
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.245             v4l2_dev.c  ...   Unsupported: Y/CbCr 4:2:2
15:19:21.245             v4l2_dev.c  ... r=0
15:19:21.245             v4l2_dev.c  ...   Unsupported: Y/CrCb 4:2:2
15:19:21.245             v4l2_dev.c  ... r=0
15:19:21.245             v4l2_dev.c  ...   Unsupported: Y/CrCb 4:2:0
15:19:21.245             v4l2_dev.c  ... r=0
15:19:21.245             v4l2_dev.c  ...   Unsupported: Y/CbCr 4:2:0
15:19:21.245             v4l2_dev.c  ... r=0
15:19:21.245             v4l2_dev.c  ...   Unsupported: Y/CrCb 4:2:0 (N-C)
15:19:21.245             v4l2_dev.c  ... r=0
15:19:21.245             v4l2_dev.c  ...   Unsupported: Y/CbCr 4:2:0 (N-C)
15:19:21.245             v4l2_dev.c  ... r=-22
15:19:21.245             v4l2_dev.c  ...   Unsupported:
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
15:19:21.245             v4l2_dev.c  ... r=-1
Using mplane plugin for output
15:19:21.246             v4l2_dev.c  ... r=0
Using mplane plugin for output
15:19:21.246             v4l2_dev.c  ... r=0
15:19:21.267             v4l2_dev.c  ... r=0
15:19:21.269             v4l2_dev.c  ... r=0
15:19:21.325             v4l2_dev.c  ... r=0
15:19:21.325             v4l2_dev.c  ...Found capture device USB Camera: USB Camera
15:19:21.325             v4l2_dev.c  ...  Enumerating formats:
15:19:21.325             v4l2_dev.c  ... r=0
15:19:21.325             v4l2_dev.c  ...   Unsupported: Motion-JPEG
15:19:21.325             v4l2_dev.c  ... r=0
15:19:21.325             v4l2_dev.c  ...   Supported: YUYV 4:2:2
15:19:21.325             v4l2_dev.c  ... r=-1
15:19:21.334             v4l2_dev.c  ... r=0
15:19:21.343             v4l2_dev.c  ... r=0
15:19:21.352             v4l2_dev.c  ... r=0
15:19:21.360             v4l2_dev.c  ... r=0
15:19:21.369             v4l2_dev.c  ... r=0
15:19:21.378             v4l2_dev.c  ... r=0
15:19:21.387             v4l2_dev.c  ... r=0
15:19:21.396             v4l2_dev.c  ... r=0
15:19:21.405             v4l2_dev.c  ... r=0
15:19:21.414             v4l2_dev.c  ... r=0
15:19:21.423             v4l2_dev.c  ... r=0
15:19:21.432             v4l2_dev.c  ... r=0
15:19:21.441             v4l2_dev.c  ... r=0
15:19:21.451             v4l2_dev.c  ... r=0
15:19:21.459             v4l2_dev.c  ... r=0
15:19:21.468             v4l2_dev.c  ... r=0
15:19:21.478             v4l2_dev.c  ... r=0
15:19:21.487             v4l2_dev.c  ... r=0
15:19:21.496             v4l2_dev.c  ... r=0
15:19:21.498             v4l2_dev.c  ...Video4Linux2 has 4 devices
FBIOGET_VSCREENINFO: Invalid argument
No such filter: '/dev/video0'
15:19:21.945              sdl_dev.c  ...SDL 2.28 initialized
15:19:21.945         colorbar_dev.c  ...Colorbar video src initialized with 2 device(s):
15:19:21.945         colorbar_dev.c  ... 0: Colorbar generator
15:19:21.945         colorbar_dev.c  ... 1: Colorbar-active
15:19:21.945         sip_endpoint.c  .Module "mod-evsub" registered
15:19:21.946         sip_endpoint.c  .Module "mod-presence" registered
15:19:21.946                evsub.c  .Event pkg "presence" registered by mod-presence
15:19:21.946         sip_endpoint.c  .Module "mod-dlg_even" registered
15:19:21.946                evsub.c  .Event pkg "dialog" registered by mod-dlg_even
15:19:21.946         sip_endpoint.c  .Module "mod-mwi" registered
15:19:21.946                evsub.c  .Event pkg "message-summary" registered by mod-mwi
15:19:21.946         sip_endpoint.c  .Module "mod-refer" registered
15:19:21.946                evsub.c  .Event pkg "refer" registered by mod-refer
15:19:21.946         sip_endpoint.c  .Module "mod-pjsua-pres" registered
15:19:21.948         sip_endpoint.c  .Module "mod-pjsua-im" registered
15:19:21.948         sip_endpoint.c  .Module "mod-pjsua-options" registered
15:19:21.948           pjsua_core.c  .1 SIP worker threads created
15:19:21.948           pjsua_core.c  .pjsua version 2.15-dev for Linux-5.10.209/aarch64/glibc-2.38 initialized
15:19:21.948           pjsua_core.c  .PJSUA state changed: CREATED --> INIT
15:19:21.948           pjsua_core.c  SIP UDP socket reachable at 192.168.137.210:5060
15:19:21.948        udp0x558090fbb0  SIP UDP transport started, published address is 192.168.137.210:5060
15:19:21.950            pjsua_acc.c  Adding account: id=<sip:1015@10.20.25.111>
15:19:21.950            pjsua_acc.c  .Account <sip:1015@10.20.25.111> added with id 0
15:19:21.950            pjsua_acc.c  .Acc 0: setting registration..
15:19:21.950            pjsua_acc.c  ..Contact for acc 0 updated: <sip:1015@192.168.137.210:5060;ob>;+sip.ice
15:19:21.950             sip_util.c  ..Request msg REGISTER/cseq=28350 (tdta0x5580914e90) created.
15:19:21.951        tsx0x5580916e50  ...Transaction created for Request msg REGISTER/cseq=28351 (tdta0x5580914e90)
15:19:21.951        tsx0x5580916e50  ..Sending Request msg REGISTER/cseq=28351 (tdta0x5580914e90) in state Null
15:19:21.951          sip_resolve.c  ...Target '10.20.25.111:0' type=Unspecified resolved to '10.20.25.111:5060' type=UDP (UDP transport)
15:19:21.951             sip_util.c  ...Resorting target addresses based on IPv4 preference
15:19:21.951           pjsua_core.c  ...TX 509 bytes Request msg REGISTER/cseq=28351 (tdta0x5580914e90) to UDP 10.20.25.111:5060:
```
as you can see from the log,  in my situation, the r is just -22 when get in error. So I rechanged the pr with the diff
``` diff
diff --git a/pjmedia/src/pjmedia-videodev/v4l2_dev.c b/pjmedia/src/pjmedia-videodev/v4l2_dev.c
index 881e6c220..170970c8c 100644
--- a/pjmedia/src/pjmedia-videodev/v4l2_dev.c
+++ b/pjmedia/src/pjmedia-videodev/v4l2_dev.c
@@ -205,7 +205,7 @@ static pj_status_t xioctl(int fh, int request, void *arg)
         r = v4l2_ioctl(fh, request, arg);
     } while (r==-1 && c++<RETRY && ((errno==EINTR) || (errno==EAGAIN)));
 
-    return (r == -1) ? pj_get_os_error() : PJ_SUCCESS;
+    return (r == -1) ? pj_get_os_error() : (r < 0 ? r :PJ_SUCCESS);
 }
 
 static vid4lin_fmt_map* get_v4l2_format_info(pjmedia_format_id id)

```

sauwming commented:
`r` is actually not directly translatable to `pj_status_t`, so perhaps just change @nanangizz 's suggested fix to:
```
    if (r != 0) {
        pj_status_t status = pj_get_os_error();
        return (status==PJ_SUCCESS)? PJ_EUNKOWN : status;
    }
```

iysheng commented:
> `r` is actually not directly translatable to `pj_status_t`, so perhaps just change @nanangizz 's suggested fix to:
> 
> ```
>     if (r != 0) {
>         pj_status_t status = pj_get_os_error();
>         return (status==PJ_SUCCESS)? PJ_EUNKOWN : status;
>     }
> ```

as you can see -22 is stand err code, is there any posssible to export a function like
```
inline (PJ_DEF)(pj_status_t) pj_convert_os_error(int err)
{
    return PJ_STATUS_FROM_OS(err);
}
```
in ``include/pj/errno.h`` file, to convert the err code ``r`` to pj_status_t type?

sauwming commented:
You don't need to create another wrapper, just use `PJ_STATUS_FROM_OS()` directly if you need it.

iysheng commented:
> You don't need to create another wrapper, just use `PJ_STATUS_FROM_OS()` directly if you need it.

good, I rechanged the pr, please recheck whether this modified can solve this **``r`` is actually not directly translatable to ``pj_status_t``**

sauwming commented:
* r is negative, so it should be: `PJ_STATUS_FROM_OS(-r)`?
* From your log, device `rkisp_selfpath` doesn't seem to return error, so we need to handle this:
```
Found capture device rkisp_selfpath
15:19:21.243             v4l2_dev.c  ...  Enumerating formats:
...
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ...   Unsupported: 16-bit RGB 5-6-5
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
15:19:21.244             v4l2_dev.c  ... r=0
...
* Also check if it's possible that `r == -1` but `errno == 0`.

iysheng commented:
> * r is negative, so it should be: `PJ_STATUS_FROM_OS(-r)`?
> * From your log, device `rkisp_selfpath` doesn't seem to return error, so we need to handle this:
> 
> ```
> Found capture device rkisp_selfpath
> 15:19:21.243             v4l2_dev.c  ...  Enumerating formats:
> ...
> 15:19:21.244             v4l2_dev.c  ... r=0
> 15:19:21.244             v4l2_dev.c  ...   Unsupported: 16-bit RGB 5-6-5
> 15:19:21.244             v4l2_dev.c  ... r=0
> 15:19:21.244             v4l2_dev.c  ... r=0
> 15:19:21.244             v4l2_dev.c  ... r=0
> ...
> ```
> 
> * Also check if it's possible that `r == -1` but `errno == 0`.

1. ok , I will change this.
2. after I run the code, the full log as
```
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...Found capture device rkisp_selfpath
00:06:40.844             v4l2_dev.c  ...  Enumerating formats:
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Supported: UYVY 4:2:2
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: Y/CbCr 4:2:2
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: Y/CrCb 4:2:2
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: Y/CrCb 4:2:0
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: Y/CbCr 4:2:0
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: 8-bit Greyscale
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: 32-bit BGRX 8-8-8-8
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: 16-bit RGB 5-6-5
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: 24-bit RGB 8-8-8
00:06:40.844             v4l2_dev.c  ... r=-22
00:06:40.844             v4l2_dev.c  ...   status: 119978
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ... r=0
``` 
as you can see -22 returned final, the log previous maybe lost something, I used tmux to log this before :sweat_smile: 
3. I thought when ret = -1, the errno code will be install errno in most scence.

iysheng commented:
> * r is negative, so it should be: `PJ_STATUS_FROM_OS(-r)`?
> * From your log, device `rkisp_selfpath` doesn't seem to return error, so we need to handle this:
> 
> ```
> Found capture device rkisp_selfpath
> 15:19:21.243             v4l2_dev.c  ...  Enumerating formats:
> ...
> 15:19:21.244             v4l2_dev.c  ... r=0
> 15:19:21.244             v4l2_dev.c  ...   Unsupported: 16-bit RGB 5-6-5
> 15:19:21.244             v4l2_dev.c  ... r=0
> 15:19:21.244             v4l2_dev.c  ... r=0
> 15:19:21.244             v4l2_dev.c  ... r=0
> ...
> ```
> 
> * Also check if it's possible that `r == -1` but `errno == 0`.

1. ok, I changed this to -r as you advice.
2. I rerun the app, the full log as
```
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...Found capture device rkisp_selfpath
00:06:40.844             v4l2_dev.c  ...  Enumerating formats:
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Supported: UYVY 4:2:2
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: Y/CbCr 4:2:2
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: Y/CrCb 4:2:2
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: Y/CrCb 4:2:0
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: Y/CbCr 4:2:0
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: 8-bit Greyscale
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: 32-bit BGRX 8-8-8-8
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: 16-bit RGB 5-6-5
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ...   status: 0
00:06:40.844             v4l2_dev.c  ...   Unsupported: 24-bit RGB 8-8-8
00:06:40.844             v4l2_dev.c  ... r=-22
00:06:40.844             v4l2_dev.c  ...   status: 119978
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ... r=0
00:06:40.844             v4l2_dev.c  ... r=0
``` 
the previous log maybe lost some info, I used tmux to log these code, :sweat_smile: as you can see final return is -22 here in  **capture device rkisp_selfpath**
3. I though when ret == -1, the errno code saveed the err code in most scene. But if ret is negative other than -1, it will be the real err code most times under good programming standards.
